### PR TITLE
[FLINK-2170] [connectors] Add ORC connector for TableSource

### PR DESCRIPTION
**Pull request by uybhatti**
## What is the purpose of the change
Currently, we can't read data from ORC files. In this PR, we added the support to load data from ORC files to Table Source.

## Brief change log
  - RowOrcInputFormat, OrcUtils and OrcTableSource classes implement the above functionality. Also, OrcTableSource implement the ProjectableTableSource and FilterableTableSource interfaces.
  - For Optimisation, reading from ORC file is done in batch instead of single row at a time.

## Verifying this change
This change added tests and can be verified as follows:
  - RowOrcInputFormatTest to verify that reading for different datatypes, nested data types and projection is correct.
  - OrcTableSourceTest and OrcTableSourceITCase are used to verify that loading of ORC data into TableSource is correct.
  
## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **yes**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`:  **no**
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): **no**
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: **no**

## Documentation

  - Does this pull request introduce a new feature? **yes**
  - If yes, how is the feature documented? **not documented**

### CODE REVIEW

File: flink-connectors/flink-orc/pom.xml [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+			<artifactId>orc-core</artifactId>
+			<version>1.4.0</version>
+		</dependency>
+
+		<!-- test dependencies -->
+
+		<dependency>
+			<groupId>org.apache.flink</groupId>
+			<artifactId>flink-test-utils-junit</artifactId>
+			<version>${project.version}</version>
```
Comment (haohui) [inline flink-connectors/flink-orc/pom.xml]: There should be no need to specify `${project.version}` here as it has been specified in the root level pom.xml

File: flink-connectors/flink-orc/pom.xml [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+			<artifactId>flink-core</artifactId>
+			<version>${project.version}</version>
+			<scope>test</scope>
+			<type>test-jar</type>
+		</dependency>
+	</dependencies>
+
+	<build>
+
+		<pluginManagement>
```
Comment (haohui) [inline flink-connectors/flink-orc/pom.xml]: Looks like unnecessary changes.

File: flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcTableSource.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+		this.orcSchema = orcSchema;
+		this.orcConfig = orcConfig;
+
+		this.typeInfo = (RowTypeInfo) OrcUtils.schemaToTypeInfo(this.orcSchema);
+
+	}
+
+	@Override
+	public DataSet<Row> getDataSet(ExecutionEnvironment execEnv) {
+
```
Comment (haohui) [inline flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcTableSource.java]: nit: additional empty line.

File: flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcTableSource.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+ * String schema = "struct<col1:boolean,col2:tinyint,col3:smallint,col4:int>"
+ * OrcTableSource orcSrc = new OrcTableSource(path, schema);
+ * tEnv.registerTableSource("orcTable", orcSrc);
+ * Table res = tableEnv.sql("SELECT * FROM orcTable");
+ * }
+ * </pre>
+ */
+public class OrcTableSource implements BatchTableSource<Row>, ProjectableTableSource<Row> {
+
+	private String path;
```
Comment (haohui) [inline flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcTableSource.java]: All of the files can be final.

File: flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcTableSource.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+	private Configuration orcConfig;
+	private int[] fieldMapping;
+
+	/**
+	 * The ORC file path and schema.
+	 *
+	 * @param path      the path of orc file
+	 * @param orcSchema schema of orc file
+	 */
+	public OrcTableSource(String path, String orcSchema) {
```
Comment (haohui) [inline flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcTableSource.java]: Create a new `Configuration` instance is very expensive. It makes sense to force the users to pass it in.

File: flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/RowOrcInputFormat.java [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+
+		this.fieldMapping = new int[this.schema.getChildren().size()];
+		for (int i = 0; i < fieldMapping.length; i++) {
+			this.fieldMapping[i] = i;
+		}
+
+	}
+
+	public void setFieldMapping(int[] fieldMapping) {
+		this.fieldMapping = fieldMapping;
```
Comment (haohui) [inline flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/RowOrcInputFormat.java]: This function can be refactored into a common function.

File: flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/RowOrcInputFormat.java [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+	}
+
+	@Override
+	public void open(FileInputSplit fileSplit) throws IOException {
+
+		this.currentSplit = fileSplit;
+		Preconditions.checkArgument(this.splitStart == 0, "ORC files must be read from the start.");
+
+		if (LOG.isDebugEnabled()) {
+			LOG.debug("Opening ORC file " + fileSplit.getPath());
```
Comment (haohui) [inline flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/RowOrcInputFormat.java]: As you're using slf4j, you can write something like 
```
LOG.debug("Opening ORC file {}", fileSplit.getPath());
```

and avoid the if statement.

File: flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/RowOrcInputFormat.java [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+		Reader.Options options = orcReader.options()
+			.range(fileSplit.getStart(), fileSplit.getLength())
+			.useZeroCopy(OrcConf.USE_ZEROCOPY.getBoolean(config))
+			.skipCorruptRecords(OrcConf.SKIP_CORRUPT_DATA.getBoolean(config))
+			.tolerateMissingSchema(OrcConf.TOLERATE_MISSING_SCHEMA.getBoolean(config));
+
+		options.include(computeProjectionMask());
+
+		// check that schema of file is as expected
+		if (!this.schema.equals(orcReader.getSchema())) {
```
Comment (haohui) [inline flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/RowOrcInputFormat.java]: It is quite common that the files actually contain different schema due to schema evolutions.

I think it is okay to leave it here as-is for now, but we definitely need a follow-up task to address it.

File: flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/RowOrcInputFormat.java [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+ * For Optimization, reading is done in batch instead of a single row.
+ */
+public class RowOrcInputFormat
+	extends FileInputFormat<Row>
+	implements ResultTypeQueryable<Row> {
+
+	private static final Logger LOG = LoggerFactory.getLogger(RowOrcInputFormat.class);
+	private static final int BATCH_SIZE = 1024;
+
+	private org.apache.hadoop.conf.Configuration config;
```
Comment (haohui) [inline flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/RowOrcInputFormat.java]: It makes sense to have a consistent naming of the field.

Comment (haohui) [reply flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/RowOrcInputFormat.java]: s/org.apache.hadoop.conf.//

File: flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/RowOrcInputFormat.java [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+/**
+ * InputFormat to read ORC data.
+ * For Optimization, reading is done in batch instead of a single row.
+ */
+public class RowOrcInputFormat
+	extends FileInputFormat<Row>
+	implements ResultTypeQueryable<Row> {
+
+	private static final Logger LOG = LoggerFactory.getLogger(RowOrcInputFormat.class);
+	private static final int BATCH_SIZE = 1024;
```
Comment (haohui) [inline flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/RowOrcInputFormat.java]: The size of the batch needs to be configurable.

File: flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcUtils.java [hunk 11]
```diff
@@ -0,0 +1,2229 @@
...
+import java.sql.Date;
+import java.sql.Timestamp;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+
+/**
+ * A class that provides utility methods for orc file reading.
+ */
+public class OrcUtils {
```
Comment (haohui) [inline flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcUtils.java]: The class can be package-local

File: flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcUtils.java [hunk 12]
```diff
@@ -0,0 +1,2229 @@
...
+				}
+				break;
+			case UNION:
+				throw new UnsupportedOperationException("UNION type not supported yet");
+			default:
+				throw new IllegalArgumentException("Unknown type " + schema);
+		}
+	}
+
+	private static void readNonNullBooleanColumn(Object[] rows, int fieldIdx, LongColumnVector vector, long[] lengthVector, int childCount) {
```
Comment (haohui) [inline flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcUtils.java]: There are significant duplications of code in these helper methods. It might need some work of refactoring.

File: flink-connectors/flink-orc/pom.xml [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+	<artifactId>flink-orc_${scala.binary.version}</artifactId>
+	<name>flink-orc</name>
+
+	<packaging>jar</packaging>
+
+	<dependencies>
+
+		<!-- core dependencies -->
+
+		<dependency>
```
Comment (haohui) [inline flink-connectors/flink-orc/pom.xml]: `${project.version}` are unnecessary as it has been specified at the root level of pom.xml

File: flink-connectors/flink-orc/pom.xml (hunk 2 above)
Comment (haohui) [inline flink-connectors/flink-orc/pom.xml]: Unnecessary changes.

File: flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcTableSource.java (hunk 4 above)
Comment (haohui) [inline flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcTableSource.java]: Mark the fields as final

File: flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcTableSource.java (hunk 5 above)
Comment (haohui) [inline flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcTableSource.java]: Creating a new instance of `Configuration` is expensive. I would force the users to pass it in.

File: flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcTableSource.java [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+
+	@Override
+	public TableSource<Row> projectFields(int[] fields) {
+
+		OrcTableSource copy = new OrcTableSource(path, orcSchema, orcConfig);
+
+		// set field mapping
+		copy.fieldMapping = fields;
+
+		// adapt TypeInfo
```
Comment (haohui) [inline flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcTableSource.java]: This function can be refactored into a common function.

File: flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/RowOrcInputFormat.java [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+
+import static org.apache.flink.orc.OrcUtils.fillRows;
+
+/**
+ * InputFormat to read ORC data.
+ * For Optimization, reading is done in batch instead of a single row.
+ */
+public class RowOrcInputFormat
```
Comment (haohui) [inline flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/RowOrcInputFormat.java]: Can be a package local class.

File: flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/RowOrcInputFormat.java (hunk 10 above)
Comment (haohui) [inline flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/RowOrcInputFormat.java]: The size of the batch needs to be configurable.

File: flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/RowOrcInputFormat.java (hunk 9 above)
Comment (haohui) [inline flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/RowOrcInputFormat.java]: Mark fields as final.

s/org.apache.hadoop.conf.//g

File: flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/RowOrcInputFormat.java [hunk 16]
```diff
@@ -0,0 +1,241 @@
...
+		}
+	}
+
+	@Override
+	public void open(FileInputSplit fileSplit) throws IOException {
+
+		this.currentSplit = fileSplit;
+		Preconditions.checkArgument(this.splitStart == 0, "ORC files must be read from the start.");
+
+		if (LOG.isDebugEnabled()) {
```
Comment (haohui) [inline flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/RowOrcInputFormat.java]: slf4j enables you to do something like `LOG.debug("Opening ORC file {}", fileSplit.getPath());` without the if statement.

File: flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/RowOrcInputFormat.java (hunk 9 above)
Comment (haohui) [reply flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/RowOrcInputFormat.java]: It also makes sense to standardize the naming of the configuration (e.g., conf)

File: flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/RowOrcInputFormat.java (hunk 8 above)
Comment (haohui) [inline flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/RowOrcInputFormat.java]: It is quite common that the files have different schemas than the one specified due to schema evolution.

It might make sense to change the error message to UnsupportedOperationException and files a follow-up jira to implement it.

File: flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcUtils.java [hunk 17]
```diff
@@ -0,0 +1,2229 @@
...
+public class OrcUtils {
+
+	/**
+	 * Convert ORC schema types to Flink types.
+	 *
+	 * @param schema schema of orc file
+	 *
+	 */
+	public static TypeInformation schemaToTypeInfo(TypeDescription schema) {
+		switch (schema.getCategory()) {
```
Comment (haohui) [inline flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcUtils.java]: For clarity it can be separated into two steps. You can do something like:

```
TypeInformation ty = TYPE_MAP.get(schema.getCategory())
if (ty != null) {
  return ty;
} else if (schema.getCategory() == ...) {
} ...
```

File: flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcUtils.java [hunk 18]
```diff
@@ -0,0 +1,2229 @@
...
+		int totalRowsInBatch = (int) batch.count();
+
+		List<TypeDescription> fieldTypes = schema.getChildren();
+		for (int outIdx = 0; outIdx < fieldMapping.length; outIdx++) {
+			int inIdx = fieldMapping[outIdx];
+			readField(rows, outIdx, fieldTypes.get(inIdx), batch.cols[inIdx], null, Math.min((int) totalRowsInBatch, rows.length));
+		}
+	}
+
+	private static void readField(Object[] rows, int fieldIdx, TypeDescription schema, ColumnVector vector, long[] lengthVector, int childCount) {
```
Comment (haohui) [inline flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcUtils.java]: There are significant of duplications across these helper functions. It requires some refactoring to clean things up.

File: flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcUtils.java (hunk 12 above)
Comment (uybhatti) [reply flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcUtils.java]: Hi @haohui, Yes true there is a lot of code duplication in helper methods, but I think it's required as I want to prefer performance over code lines/code quality. For example, I can make one method for boolean, byte, short, int, long, double and float. But then for every element in a column, first I have to check the type and then have to cast it according to this type. Which is not efficient.
Also, I am not checking `if` condition in a `for` loop, rather I am running `for` loop depending upon the `if` condition. Which is also a reason for code duplication.
Thanks for your feedback, please give me more feedback on these points.
Thanks, Usman

**Comment by fpompermaier** [comment]
Any update on this? reading ORC data would be a very nice feature..
